### PR TITLE
[BEAM-1225]: Utility class to convert PCollection<T> to PCollection<String>

### PR DESCRIPTION
**Pull request by vikkyrk**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by vikkyrk** [comment]
R: @dhalperi @davorbonaci I find this to be a useful transform for users. Thoughts?. Also should this be in tranforms package or somewhere else?

**Comment by davorbonaci** [comment]
JIRA? Discussion?

R: @dhalperi

**Comment by jbonofre** [comment]
This sounds related to data format extension discussed on the mailing list and with small PoC there:

https://github.com/jbonofre/beam/tree/DATAFORMAT

**Comment by jbonofre** [comment]
Not sure it makes sense to have an "isolated" transform on the core. I would better package all related data format/converters transforms in an unique extension (as proposed). Thoughts ?

**Comment by dhalperi** [comment]
I think there was a long discussion on this issue in the Beam mailing list: https://lists.apache.org/thread.html/CAOdmQRguKoDgVHfX7FBFeikiX8ezOjTGPjcews6cR1kNuu84pA@mail.gmail.com

Suggest joining the discussion there. CC: @eljefe6a

**Comment by vikkyrk** [comment]
Thanks all, I wasn't aware of the thread. I will bring this up there.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+package org.apache.beam.sdk.transforms;
+
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * A {@link PTransform} that converts a {@link PCollection} of type {@code T} to a
+ * {@link PCollection} of type {@code String}.
+ */
+public class ToString<T> extends PTransform<PCollection<T>, PCollection<String>> {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java]: `final`. Make it not extend PTransform, but stand alone.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * Returns a {@link ToString} transform.
+   *
+   * @param <T> the type of the input {@code PCollection}.
+   */
+  public static <T> ToString<T> create() {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java]: Make this type signature return `PTransform<T, String>`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    return new ToString<>();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java]: return new ToString.Default<>();

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  private ToString() {
+  }
+
+  @Override
+  public PCollection<String> expand(PCollection<T> input) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ToString.java]: move this into a `private static final class Default extends PTransform<T, String>` inner class that does what you think it does.

This way we can have a `.kv(String formatString)` later, for example, depending on what Jesse and co decide.

**Comment by bjchambers** [comment]
+1 to the comments, although it would be good if we could come up with a name that is more descriptive than `Default` which states nothing about what the transform is actually doing. Maybe `ElementToString` or something like that?

**Comment by vikkyrk** [comment]
`Default` makes sense to me in the sense that `toString` method is a default string conversion method in Java and would works for all types.

I would have preferred `create` and used it for kv and iterables too but type erasure wouldn't allow it. So I am leaning towards:
`ToString.default(), ToString.kv(String delimiter), ToString.iterable(String delimiter) and so on...`

**Comment by vikkyrk** [comment]
Except that `default` is a keyword in Java8. So `ToString(), ToString.kv(String delimiter), ToString.iterable(String delimiter) and so on...` instead?

**Comment by bjchambers** [comment]
I don't think we can do `ToString()`, so maybe:

`ToString.element()` (calls toString on each element)
`ToString.kv(String fmt)` (for generality, may be better to take a format string rather than a delimiter)
`ToString.iterable(String delimiter)`

**Comment by vikkyrk** [comment]
Done

**Pull request merged**